src/coreclr/vm/methodtable.cpp

```diff
@@ -624,8 +516,13 @@ BOOL MethodTable::HasSameTypeDefAs(MethodTable *pMT)
     if (GetTypeDefRid() != pMT->GetTypeDefRid())
         return FALSE;
 
-    if (GetCanonicalMethodTable() == pMT->GetCanonicalMethodTable())
-        return TRUE;
+    // Compare CanonicalMethodTables ahead of comparing Module
+    // Array MethodTables result of GetModule cannot be used for type comparison
+    // as the Rank is not part of the rid, and the Module may not
+    // match even if the MethodTable should return a match. So compare
+    // CanonicalMethodTable instead.
+    if (IsArray())
+        return GetCanonicalMethodTable() == pMT->GetCanonicalMethodTable();
```

Comment:
This looks like a bug in the existing implementation. It does not make sense for `object[]` and `string[]` to have same RID, but `object[]` and `int[]` to not have the same RID. I think this can be simplified to:

```
// optimize for the negative case where we expect RID mismatch
unsigned rid = GetTypeDefRid();
if (rid != pMT->GetTypeDefRid())
    return FALSE;

// Types without RIDs are unrelated to each other. This case is taken for arrays. 
if (rid == 0)
    return FALSE;

return (GetModule() == pMT->GetModule());
```